homeassistant/components/fritz/sensor.py

```diff
@@ -0,0 +1,141 @@
+"""AVM FRITZ!Box binary sensors."""
+from __future__ import annotations
+
+import datetime
+import logging
+
+from fritzconnection.core.exceptions import FritzConnectionException
+
+from homeassistant.components.binary_sensor import BinarySensorEntity
+from homeassistant.config_entries import ConfigEntry
+from homeassistant.const import DEVICE_CLASS_TIMESTAMP
+from homeassistant.core import HomeAssistant
+from homeassistant.util.dt import utcnow
+
+from .common import FritzBoxBaseEntity, FritzBoxTools
+from .const import DOMAIN, UPTIME_DEVIATION
+
+_LOGGER = logging.getLogger(__name__)
+
+
+def _retrieve_uptime_state(status, last_value):
+    """Return uptime from device."""
+    delta_uptime = utcnow() - datetime.timedelta(seconds=status.uptime)
+
+    if (
+        not last_value
+        or abs(
+            (delta_uptime - datetime.datetime.fromisoformat(last_value)).total_seconds()
+        )
+        > UPTIME_DEVIATION
+    ):
+        return delta_uptime.replace(microsecond=0).isoformat()
+
+    return last_value
+
+
+def _retrieve_external_ip_state(status, last_value):
+    """Return external ip from device."""
+    return status.external_ip
+
+
+SENSOR_NAME = 0
+SENSOR_DEVICE_CLASS = 1
+SENSOR_ICON = 2
+SENSOR_STATE_PROVIDER = 3
+
+# sensor_type: [name, device_class, icon, state_provider]
+SENSOR_DATA = {
+    "external_ip": [
+        "External IP",
+        None,
+        "mdi:earth",
+        _retrieve_external_ip_state,
+    ],
+    "uptime": ["Uptime", DEVICE_CLASS_TIMESTAMP, None, _retrieve_uptime_state],
+}
+
+
+async def async_setup_entry(
+    hass: HomeAssistant, entry: ConfigEntry, async_add_entities
+) -> None:
+    """Set up entry."""
+    _LOGGER.debug("Setting up FRITZ!Box sensors")
+    fritzbox_tools = hass.data[DOMAIN][entry.entry_id]
+
+    if "WANIPConn1" not in fritzbox_tools.connection.services:
+        # Only routers are supported at the moment
+        return
+
+    for sensor_type in SENSOR_DATA:
+        async_add_entities(
+            [FritzBoxSensor(fritzbox_tools, entry.title, sensor_type)],
+            True,
+        )
+
+
+class FritzBoxSensor(FritzBoxBaseEntity, BinarySensorEntity):
+    """Define FRITZ!Box connectivity class."""
+
+    def __init__(
+        self, fritzbox_tools: FritzBoxTools, device_friendlyname: str, sensor_type: str
```

Comment:
Please separate words with underscore in parameter and variable names. `friendly_name`